Add missing dependencies and extra info to optimizer step docs (#15054)
ethereum/solidity: docs/internals/optimizer.rst

into a certain normal form that is easier to work with. This normal
form is kept during the rest of the optimization process.

Some of the steps in this group should be considered prerequisites for all other components.
Currently these are:

> - :ref:`disambiguator`

Reviewer: above we write `Some steps depend on properties ensured by ...`, which does not list disambiguator but the rest of the four. For consistency, I'd also add it there (line 349, for some reason, gh does not want me to add a review comment there)

- :ref:`function-hoister`
- :ref:`function-grouper`

> - :ref:`for-loop-init-rewriter`

Reviewer: Can we reorder the detailed description of the steps so that it reflects this order? Right now it is `Disambiguator`, `FunctionHoister`, `FunctionGrouper`, `ForLoopConditionIntoBody`, and only then `ForLoopInitRewriter`

The benefit of this stage is that function definitions can be looked up more easily
and functions can be optimized in isolation without having to traverse the AST completely.

> Prerequisites: Disambiguator.

Author: Note that not all steps list disambiguator as a prerequisite even though its own description above says that all of them depend on it.

BTW, not sure if listing prerequisites is all that useful since we have very few hard prerequisites and those are hard-coded anyway. Maybe it would be fine to remove this info. For now though we do list that info so I amended it for steps where it was available in the docstrings.

This transformation can also be useful when paired with LoopInvariantCodeMotion, since
invariants in the loop-invariant conditions can then be taken outside the loop.

Loops that already have a literal constant as iteration condition are not transformed.

To avoid unnecessary rewriting, it is recommended to run this step after StructuralSimplifier.

> May destroy the :ref:`expression-split <expression-splitter>` form.

Author: I inferred the info about which steps preserve the expression-split and SSA forms from the current docs. Please double check that I did not make a mistake here. Since most steps seem to preserve the forms, I used the convention of stating it only if the step can destroy it.

For example this step (`ForLoopConditionIntoBody`) may destroy expression-split form because it would transform something like this:

```yul
for { ... } iszero(a) { ... } {
    ...
}
```

into this:
```yul
for { ... } 1 { ... } {
    if iszero(iszero(a)) { break }
}
```

`ExpressionSimplifier` also seems to mostly preserve it but also has some rules that don't. For example the `exp` rule will transform this:

```yul
let _0 := 0
let a := calldataload(_0)
let b := not(_0)
let c := exp(b, a)
sstore(a, c)
```

into this:
```yul
let _0 := 0
let a := calldataload(_0)
let b := 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
let c := sub(iszero(and(a, 1)), and(a, 1))
sstore(a, c)
```

Reviewer: I would phrase it a bit differently, what do you think about

```suggestion
The :ref:`expression-split <expression-splitter>` form is not guaranteed to be preserved under this step.
```

This transformation can also be useful when paired with LoopInvariantCodeMotion, since
invariants in the loop-invariant conditions can then be taken outside the loop.

> Loops that already have a literal constant as iteration condition are not transformed.

Reviewer: ```suggestion
Loops that already have a literal constant or an individual identifier as iteration condition are not transformed.
```

> To avoid unnecessary rewriting, it is recommended to run this step after StructuralSimplifier.

Reviewer: How so, wouldn't the structural simplifier just remove for loops that have a `false` condition with its initialization? These cases would be ignored by the `ForLoopConditionIntoBody` step anyways

UnusedAssignEliminator will then be able to fully eliminate such
variables.

The longer the values of variables are available, the higher the chance of an expression being replaceable.
For this reason the step could potentially be more effective when used on code in pseudo-SSA form.

> This property, however, is not easily destroyed once established, and in most cases it is enough

Reviewer: It is not quite clear from the text what "this" property should be (pseudo ssa form vs long-livedness of variables' values), we could phrase it as 

```suggestion
The longevity of variable values' availability, however, is mostly preserved once established, and therefore in most cases it is enough
```

if SSATransform was used once at any point before running the step.

> The difference may also not be very significant in practice.

Reviewer: The difference of what vs what? If it is running SSA immediately beforehand vs just somewhere before, I suggest to drop this sentence

During the matching procedure, it resolves variables to their currently
assigned expressions to be able to match more deeply nested patterns
even when the code is in pseudo-SSA form.
In fact, running CommonSubexpressionEliminator before helps this component track equivalent subexpressions.

Some of the patterns like ``X - X -> 0`` can only be applied as long
as the expression ``X`` is movable, because otherwise it would remove its potential side-effects.
Since variable references are always movable, even if their current
value might not be, the ExpressionSimplifier is again more powerful
in :ref:`expression-split <expression-splitter>` or :ref:`pseudo-SSA <ssa-transform>` form.

> Some of the simplifications destroy the :ref:`expression-split <expression-splitter>` form.

Reviewer: As above I would phrase it as the expression split form not being guaranteed to be preserved

.. _literal-rematerialiser:

LiteralRematerialiser
^^^^^^^^^^^^^^^^^^^^^

If a variable referenced in an expression is known to have a literal value, replaces that use of
the variable with the literal.

This is mostly used so that other components do not have to rely on the DataflowAnalyzer.

While the step destroys the :ref:`expression-split <expression-splitter>` form in the strict sense,
it does not introduce any nested function calls, which may be good enough for many steps that require

> this property.

Author: Also clarified this for `LiteralRematerialiser`, based on our discussion on the chat:

> You also don't need to consider the LiteralRematerializer to destroy expression-split form - it doesn't do so in a relevant way - even though it'd be interesting and to check if that makes any difference. I.e. if we find any case, in which this changes, and it's only due to the LiteralRematerializer, then it's likely ok (which we could then verify directly)

Reviewer: Does any step actually require the ast to be in expression split form as prerequisite? Isn't it rather that some steps run potentially more efficiently if it is fulfilled? In that case I would phrase it more like

```suggestion
While the step destroys the :ref:`expression-split <expression-splitter>` form in the strict sense,
it does not introduce any nested function calls, which may still be good enough for many steps that require
this property to function efficiently.
```

Reviewer: Impacts at least `EqualStoreEliminator`

Optimisation stage that replaces expressions of type ``sload(x)`` and ``mload(x)`` by the value
currently stored in storage resp. memory, if known.

Also evaluates simple ``keccak256(a, c)`` when the value at memory location ``a`` is known and ``c``
is a constant ``<= 32``.

> Works best if the code is in SSA form.

Reviewer: but this would also invalidate the (strict) expression split form, I think

Reviewer: impacts effectiveness of `EqualStoreEliminator`

will not change and the CommonSubexpressionEliminator re-uses exactly the same
variable if the value is known to be the same.

Works best on code without literal arguments, which is the case when the code is in the

> :ref:`expression-split <expression-splitter>` form.

Author: I based this addition on the docstrings, which says this: https://github.com/ethereum/solidity/blob/fb11553a650f6e07299354a867fc80e1c1ba3c0b/libyul/optimiser/EqualStoreEliminator.h#L35

Since SSATransform on its own does nothing about literal arguments, this basically means that that ExpressionSplitter is recommended before, right?

Reviewer: it uses `isSimpleStore` of the `DataFlowAnalyzer`, which checks if the key and value arguments are identifiers, so indeed literals as arguments would fly under the radar and expression split form is recommended. This is also a place where destroying the strict expression split form would hurt (f ex in load resolver or literal rematerializer).

This component uses the DataflowAnalyzer.

> LiteralRematerialiser should be run before this step.

Reviewer: it destroys the expression-split form e.g. if the condition of an `if`-statement is a function call but the body is empty.
`if (iszero(a)) {}` -> `pop(iszero(a))`

This component uses the DataflowAnalyzer.

LiteralRematerialiser should be run before this step.

Prerequisites: Disambiguator.

> Important: Can only be used on EVM code.

Author: Not sure about usefulness of this either now that we only have an EVM backend, but since other steps already list this kind of info (e.g. `Important: Introduces EVM opcodes and thus can only be used on EVM code for now` for ControlFlowSimplifier) I added it where it was available for now.

Reviewer 2: The code only has that, because we haven't removed all of it when removing the EWASM parts :-).

Author: I mean, this still info makes some sense unless we know for sure that there will never be backends other than EVM. I suspect it's incomplete so it may be not very useful even then, but I don't know for sure so I didn't want to remove it without some confirmation.

Author: Remarks about EVM removed from the docs and also from the docstrings.

ExpressionSplitter and SSATransform should be run upfront to obtain better results.

> The transformation will not move loop-invariant condition out of the condition block of the loop.

Reviewer: ```suggestion
The transformation will not move loop-invariant conditions out of the condition block of the loop.
```

The inliner should be run afterwards to make sure that all references to ``f2`` are replaced by

> ``f``.

Author: I'm assuming this means `ExpressionInliner`, because it inlines functions of exactly this form that this step introduces. Is that correct?

Reviewer 2: No, that's just coincidence here - if you have a function with three return arguments and only one of them is unused, so two remain, only the full inliner can inline this.

Author: Thanks. That's very useful to know. I'll update the docs here.

Author: Updated to `FullInliner`.

  Note that indirect recursive calls (going through other functions) do not disqualify the function from inlining.
- Argument expressions passed into the call can potentially have side effects.
  This check is very broad, rejecting anything that is not simply a literal or an identifier, since
  the condition is easy to satisfy by running ExpressionSplitter beforehand.

Otherwise, if any of the following conditions is met, the call is inlined unconditionally:

- The function is tiny.
- The function is only ever called once.

In the remaining cases the decision to inline or not depends on the size of the function.
The size used here is not the actual bytecode size (which is not known to the optimizer), but rather
a score calculated at the Yul level, which approximates that size.
Inlining is performed if the size does not exceed a predefined threshold, which is a little higher
if the inlining is aggressive and yet higher when the call has some constant arguments.

The aggressive inlining is performed when:

- The :ref:`yul-memoryguard` is present.
- The function is not recursive (not even indirectly).

> - The legacy Yul->EVM transform is not used (i.e. the target is not the ``homestead`` :ref:`EVM version <evm-version>`).

Author: This is a summary of what [`FullInliner::shallInline()`](https://github.com/ethereum/solidity/blob/v0.8.25/libyul/optimiser/FullInliner.cpp#L178-L238) does.

Reviewer 2: I really don't think it is good to document things into this level of detail :-). There is little merit in it, it only leads to people *relying* on some behaviour we want to be able to change and it's a maintenance hassle.

Author: This is something I needed to know multiple times in the past and each time I just looked at the code of `shallInline()`, but it's nowhere as clear as having it spelled out in the docs and I never had time to dive into details so I actually never fully understood how it really worked. For example, that the `m_noInlineFunctions` check is simply a check for functions using `leave`. Or how some indirect recursive calls may actually be subject to inlining (there are three separate checks for recursive calls and it's not obvious how they differ). This time I needed it again to figure out why a function was not being inlined in one of my snippets. It always felt like something that should have been documented *somewhere*.

I could put it the docstring or inside `FullInliner::shallInline()`, but I really don't see much problem with having it in the docs proper. It's our main resource to understand how the optimizer works, not just for users but also for us and that fully justifies some extra effort spent keeping it up to date.

At the same time I don't think the extra detail really hurts users. It does not go into tiny details of how the inliner is implemented. It just describes the heuristic, which is useful to know when you're optimizing your code. But even then it omits all the specifics that could easily change, like numbers or details of the metric. There are step descriptions that are much more detailed than that. So if you're worried about users relying on them - we already have that problem. I don't think it's a good reason to avoid documenting things. I think that we in general lean too hard on that and I think it's one of the reasons why the code is so hard to get into.

In any case, I feel very strongly that the inliner heuristic should be documented.
